examples/pixelcnn/train.py

```diff
@@ -0,0 +1,280 @@
+  metrics = jax.lax.pmean(metrics, 'batch')
+  return metrics
+
+
+def train_step(optimizer, state, batch, prng_key, learning_rate_fn):
+  """Perform a single training step."""
+  def loss_fn(model):
+    """loss function used for training."""
+    # TODO(j-towns): do we need state for this model?
+    with flax.nn.stateful(state) as new_state:
+      with flax.nn.stochastic(prng_key):
+        nn_out = model(batch['image'], dropout_p=FLAGS.dropout_rate)
+    loss = neg_log_likelihood_loss(nn_out, batch['image'])
+    return loss, new_state
+
+  step = optimizer.state.step
+  lr = learning_rate_fn(step)
+  grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+  (loss, new_state), grad = grad_fn(optimizer.target)
+  grad = jax.lax.pmean(grad, 'batch')
+  new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+
+  metrics = {'loss': jax.lax.pmean(loss, 'batch')}
+  metrics['learning_rate'] = lr
+  return new_optimizer, new_state, metrics
+
+
+def eval_step(model, state, batch):
+  state = jax.lax.pmean(state, 'batch')
+  with flax.nn.stateful(state, mutable=False):
+    nn_out = model(batch['image'], dropout_p=0)
+  return compute_metrics(nn_out, batch['image'])
+
+
+def load_and_shard_tf_batch(xs):
+  local_device_count = jax.local_device_count()
+  def _prepare(x):
+    # Use _numpy() for zero-copy conversion between TF and NumPy.
+    x = x._numpy()  # pylint: disable=protected-access
+    return x.reshape((local_device_count, -1) + x.shape[1:])
+
+  return jax.tree_map(_prepare, xs)
+
+
+def train(model_def, model_dir, batch_size, init_batch_size, num_epochs,
+          learning_rate, decay_rate, run_seed=0):
+  """Train model."""
+  if jax.host_count() > 1:
+    raise ValueError('PixelCNN++ example should not be run on more than 1 host'
+                     ' (for now)')
+
+  summary_writer = tensorboard.SummaryWriter(model_dir)
+
+  rng = random.PRNGKey(run_seed)
+
+  if batch_size % jax.device_count() > 0:
+    raise ValueError('Batch size must be divisible by the number of devices')
+  device_batch_size = batch_size // jax.device_count()
+
+  # Load dataset
+  data_source = input_pipeline.DataSource(
+      train_batch_size=batch_size, eval_batch_size=batch_size)
+  train_ds = data_source.train_ds
+  eval_ds = data_source.eval_ds
+
+  # Create dataset batch iterators
+  train_iter = iter(train_ds)
+  eval_iter = iter(eval_ds)
+
+  # Compute steps per epoch and nb of eval steps
+  steps_per_epoch = data_source.TRAIN_IMAGES // batch_size
+  steps_per_eval = data_source.EVAL_IMAGES // batch_size
+  num_steps = steps_per_epoch * num_epochs
+
+  base_learning_rate = learning_rate
+
+  # Create the model using data-dependent initialization. Don't shard the init
+  # batch.
+  assert init_batch_size <= batch_size
+  init_batch = next(train_iter)['image']._numpy()[:init_batch_size]
+  model, state = create_model(rng, init_batch, model_def)
+  state = jax_utils.replicate(state)
+  optimizer = create_optimizer(model, base_learning_rate)
```

Comment:
Consider adding checkpointing. It's actually pretty easy. Check the ImageNet example for that.

Reply:
Thanks, done.